### PR TITLE
Non throwing directory creation function.

### PR DESCRIPTION
**Pull request by akindyakov**
Non throwing  directory creation function.

`Status` was used as a return value, as far as all function in this file operate with `Status` not `Expected`. Let's move from `Status` to `Expected` in another PR.

### CODE REVIEW

File: include/osquery/filesystem.h [hunk 1]
```diff
@@ -234,6 +234,19 @@ Status movePath(const boost::filesystem::path& from,
  */
 Status isDirectory(const boost::filesystem::path& path);
 
+/**
+ * @brief Create the directory
+ *
+ * @param path to the intended directory
+ * @param recursive - make parent directories as needed
```
Comment (guliashvili) [inline include/osquery/filesystem.h]: NIT. also create missing intermediate-level directories

File: osquery/filesystem/filesystem.cpp [hunk 2]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+    is_created = boost::filesystem::create_directory(dir_path, err);
+  }
+  if (is_created) {
+    return Status::success();
+  }
+  if (ignore_existence && isDirectory(dir_path).ok()) {
+    return Status::success();
+  }
+  auto msg = std::string{"Could not create directory "} + dir_path.string();
+  if (err) {
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: err != boost::system::errc::success

File: osquery/filesystem/filesystem.cpp [hunk 3]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+  } else {
+    is_created = boost::filesystem::create_directory(dir_path, err);
+  }
+  if (is_created) {
+    return Status::success();
+  }
+  if (ignore_existence && isDirectory(dir_path).ok()) {
+    return Status::success();
+  }
+  auto msg = std::string{"Could not create directory "} + dir_path.string();
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: quote(dir_path..) ?

File: osquery/filesystem/filesystem.cpp [hunk 4]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+  if (err) {
+    msg += ": ";
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: I think message would looks strange
Could not create directory /var/x: some error

File: osquery/filesystem/filesystem.cpp [hunk 5]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+  return Status::failure(msg);
+}
+
+Status createDirectoryRecursive(const boost::filesystem::path& dir_path) {
+  auto err = boost::system::error_code{};
+  if (boost::filesystem::create_directories(dir_path, err)) {
+    return Status::success();
+  }
+  auto msg =
+      std::string{"Could not create directory and parents"} + dir_path.string();
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: and -> or?

File: osquery/filesystem/filesystem.cpp [hunk 6]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+  }
+  auto msg = std::string{"Could not create directory "} + dir_path.string();
+  if (err) {
+    msg += ": ";
+    msg += err.message();
+  }
+  return Status::failure(msg);
+}
+
+Status createDirectoryRecursive(const boost::filesystem::path& dir_path) {
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: Why do we need separate implementation for that? createDirectory can do it anyway, if it's just for simplicity why don't we  call something like createDirectory(dir_path, true, true);

File: osquery/filesystem/filesystem.cpp [hunk 7]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+  auto err = boost::system::error_code{};
+  if (boost::filesystem::create_directories(dir_path, err)) {
+    return Status::success();
+  }
+  auto msg =
+      std::string{"Could not create directory and parents"} + dir_path.string();
+  if (err) {
+    msg += ": ";
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: same

File: osquery/filesystem/filesystem.cpp [hunk 8]
```diff
@@ -438,6 +438,44 @@ Status isDirectory(const fs::path& path) {
...
+}
+
+Status createDirectoryRecursive(const boost::filesystem::path& dir_path) {
+  auto err = boost::system::error_code{};
+  if (boost::filesystem::create_directories(dir_path, err)) {
+    return Status::success();
+  }
+  auto msg =
+      std::string{"Could not create directory and parents"} + dir_path.string();
+  if (err) {
```
Comment (guliashvili) [inline osquery/filesystem/filesystem.cpp]: same

File: osquery/filesystem/tests/filesystem_tests.cpp [hunk 9]
```diff
@@ -97,12 +97,19 @@ TEST_F(FilesystemTests, test_remove_path) {
...
   EXPECT_TRUE(removePath(test_dir));
-  EXPECT_FALSE(pathExists(test_file));
-  EXPECT_FALSE(pathExists(test_dir));
+  EXPECT_FALSE(pathExists(test_file).ok());
+  EXPECT_FALSE(pathExists(test_dir).ok());
+}
+
+TEST_F(FilesystemTests, test_remove_path_akindyakov) {
+  fs::path test_path(kTestWorkingDirectory);
+  test_path /= "akindyakov_th";
```
Comment (guliashvili) [inline osquery/filesystem/tests/filesystem_tests.cpp]: 😸

**Comment by guliashvili** [comment]
Very good. We should take care about this boost throwing problem. I'm also working on some other issue about the boost usage which kills osquery #5194

File: osquery/filesystem/filesystem.cpp (hunk 6 above)
Comment (akindyakov) [reply osquery/filesystem/filesystem.cpp]: Oh, thank you! I just forgot to delete it.

File: osquery/filesystem/filesystem.cpp (hunk 5 above)
Comment (akindyakov) [reply osquery/filesystem/filesystem.cpp]: deleted

File: osquery/filesystem/filesystem.cpp (hunk 4 above)
Comment (akindyakov) [reply osquery/filesystem/filesystem.cpp]: Added quoted around the path.  Thanks!

File: osquery/filesystem/tests/filesystem_tests.cpp (hunk 9 above)
Comment (akindyakov) [reply osquery/filesystem/tests/filesystem_tests.cpp]: whoops, thank you!

**Pull request merged**